client/signup/config/flows.js

```diff
@@ -238,11 +239,72 @@ const Flows = {
 			flow = filterDesignTypeInFlow( flow );
 		}
 
-		return flow;
+		return Flows.getABTestFilteredFlow( flowName, flow );
 	},
 
 	getFlows() {
 		return flows;
+	},
+
+	/**
+	 * Preload AB Test variations after a certain step has been completed.
+	 *
+	 * This gives the option to set the AB variation as late as possible in the
+	 * signup flow.
+	 *
+	 * Currently only the `main` flow is whitelisted.
+	 *
+	 * @param {String} flowName The current flow
+	 * @param {String} stepName The step that is being completed right now
+	 */
+	preloadABTestVariationsForStep( flowName, stepName ) {
```

Comment:
> The flow should not be modified until the AB test variation is assigned to the user.

Totally agreed. But you're calling both `flows.preloadABTestVariationsForStep()` and `flows.getFlow()` each time you load the next step. You could just as easily inject the step inside `flow.getFlow()` so you only need to worry about it in one spot... you just need to check the flowname and the stepname before assigning a variant and injecting the dynamic step, just like you're doing in `preloadABTestVariationsForStep()`. It requires passing through the stepname into `getFlow()` to make it available in `getABTestFilteredFlow()` but I think that's preferable to creating a separate method just for injecting the step, and having to maintain branching logic in two places.

One other thing, in case a user navigates back and forth through the steps several times you should check if the new step already exists before injecting it. I think that would be simpler and a little more clear than what you're doing with `.resumingFlow`.